nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ScanAttribute.java

```diff
@@ -206,36 +243,60 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
         final boolean matchAll = context.getProperty(MATCHING_CRITERIA).getValue().equals(MATCH_CRITERIA_ALL);
 
-        for (final FlowFile flowFile : flowFiles) {
-            final boolean matched = matchAll ? allMatch(flowFile, attributePattern, dictionaryTerms) : anyMatch(flowFile, attributePattern, dictionaryTerms);
-            final Relationship relationship = matched ? REL_MATCHED : REL_UNMATCHED;
+        for  (FlowFile flowFile : flowFiles) {
+            final Map<String,String> matched = (matchAll ? matchAll(flowFile, attributePattern, dictionaryTerms) : matchAny(flowFile, attributePattern, dictionaryTerms));
+            flowFile = session.putAllAttributes(flowFile, matched);
+
+            final Relationship relationship = (((matched.size() == (attributeNameMatches.size() * 3) && matchAll) || (matched.size() > 0 && !matchAll))) ? REL_MATCHED : REL_UNMATCHED;
             session.getProvenanceReporter().route(flowFile, relationship);
             session.transfer(flowFile, relationship);
             logger.info("Transferred {} to {}", new Object[]{flowFile, relationship});
         }
     }
 
-    private boolean allMatch(final FlowFile flowFile, final Pattern attributePattern, final Set<String> dictionary) {
-        for (final Map.Entry<String, String> entry : flowFile.getAttributes().entrySet()) {
-            if (attributePattern == null || attributePattern.matcher(entry.getKey()).matches()) {
-                if (!dictionary.contains(entry.getValue())) {
-                    return false;
+    private Map<String,String> matchAny(final FlowFile flowFile, final Pattern attributePattern, final Map<String,String> dictionary) {
+        Map<String,String> dictionaryTermMatches = new HashMap<String,String>();
+        attributeNameMatches = new HashSet<String>();
+
+        int hitCounter = 0;
+
+        for (final Map.Entry<String, String> attribute : flowFile.getAttributes().entrySet()) {
+            if (attributePattern == null || attributePattern.matcher(attribute.getKey()).matches()) {
+                attributeNameMatches.add(attribute.getKey());
+
+                if (dictionary.containsKey(attribute.getValue())) {
+                    hitCounter++;
+                    dictionaryTermMatches.put("dictionary.hit." + hitCounter + ".attribute", attribute.getKey());
+                    dictionaryTermMatches.put("dictionary.hit." + hitCounter + ".term", attribute.getValue());
+                    dictionaryTermMatches.put("dictionary.hit." + hitCounter + ".metadata", dictionary.get(attribute.getValue()));
```

Comment:
It looks like these statements could be refactored to a method as they are reused in the `matchAll` scenario.